crates/runtime/src/outband_fuel/linux.rs

```diff
@@ -0,0 +1,218 @@
+        // the pthread_t is initialized.
+        let info = match info {
+            Some(info) => info,
+            None => return true,
+        };
+
+        let thread_state = get_thread_state(context);
+        if IS_WASM_PC(thread_state.pc) {
+            // at this point the fuel register is defined.
+
+            // flush it into the VMRuntimeLimits.
+            *(*info.runtime_limits()).fuel_consumed.get() = thread_state.fuel;
+
+            // check if the fuel ran out and if so interrupt.
+            if thread_state.fuel > 0 {
+                raise_lib_trap(TrapCode::Interrupt);
+            }
+        }
+
+        true
+    });
+
+    if handled {
+        return;
+    }
+
+    // This signal is not for any compiled wasm code we expect, so we
+    // need to forward the signal to the next handler. If there is no
+    // next handler (SIG_IGN or SIG_DFL), then it's time to crash. To do
+    // this, we set the signal back to its original disposition and
+    // return. This will cause the faulting op to be re-executed which
+    // will crash in the normal way. If there is a next handler, call
+    // it. It will either crash synchronously, fix up the instruction
+    // so that execution can continue and return, or trigger a crash by
+    // returning the signal to it's original disposition and returning.
+    let previous = &*PREV_SIGUSR1.as_ptr();
+    if previous.sa_flags & libc::SA_SIGINFO != 0 {
+        mem::transmute::<usize, extern "C" fn(libc::c_int, *mut siginfo_t, *mut libc::c_void)>(
+            previous.sa_sigaction,
+        )(signum, siginfo, context)
+    } else if previous.sa_sigaction == libc::SIG_DFL || previous.sa_sigaction == libc::SIG_IGN {
+        libc::sigaction(signum, previous, ptr::null_mut());
+    } else {
+        mem::transmute::<usize, extern "C" fn(libc::c_int)>(previous.sa_sigaction)(signum)
+    }
+}
+
+struct ThreadState {
+    pc: usize,
+    /// The contents of the fuel register.
+    ///
+    /// Only defined if `pc` points to a code compiled from wasm.
+    fuel: i64,
+}
+
+unsafe fn get_thread_state(cx: *mut libc::c_void) -> ThreadState {
+    cfg_if::cfg_if! {
+        if #[cfg(all(target_os = "linux", target_arch = "x86_64"))] {
+            let cx = &*(cx as *const libc::ucontext_t);
+            let rip = cx.uc_mcontext.gregs[libc::REG_RIP as usize] as *const u8;
+            let r15 = cx.uc_mcontext.gregs[libc::REG_R15 as usize];
+            ThreadState {
+                pc: rip as usize,
+                fuel: r15,
+            }
+        } else if #[cfg(all(target_os = "macos", target_arch = "aarch64"))] {
+            let cx = &*(cx as *const libc::ucontext_t);
+            let pc = (*cx.uc_mcontext).__ss.__pc as *const u8;
+            let x21 = (*cx.uc_mcontext).__ss.__x[21] as i64;
+            ThreadState {
+                pc: pc as usize,
+                fuel: x21,
+            }
+        } else {
+            panic!("unsupported platform")
+        }
+    }
+}
```

Comment:
The platform-specific code here and elsewhere in this PR relies on quite some C&P.

I wondered if it would be a good idea to reshuffle the low-level code into a separate `sys` module. It will then contain the platform-specific parts for traphandlers, out-of-band fuel and whatnot.